File: 5.10/Capacity_Refinement_MultiRSU.m
function B_Final = Capacity_Refinement_MultiRSU(B_MWC, W_net, Tile_Size, dependency_matrix, C_RSU, E, X)
% Capacity_Refinement_MultiRSU - 优化版：逐次剔除低效用瓦片+依赖感知替代保留
% 核心改进：
% 1. 不清空初始缓存，仅剔除超出容量的低效用瓦片
% 2. 按效用升序逐次剔除，而非全清重选
% 3. 依赖违规时优先替换同依赖链高效用瓦片，减少连锁剔除

num_tiles = length(W_net);
B_Final = B_MWC; % 保留MWC初始缓存集，不清空
Psi = W_net ./ Tile_Size; % 效用密度（越大越优先保留）
AccumulatedSize = zeros(1, E); 

% 1. 初始化各RSU当前容量
for r = 1:E
    start_idx = (r-1)*X + 1;
    end_idx = r*X;
    AccumulatedSize(r) = sum(B_Final(start_idx:end_idx) .* Tile_Size(start_idx:end_idx));
end

% 2. 逐RSU处理容量超限：逐次剔除低效用瓦片
for r = 1:E
    C_r = C_RSU(r);
    start_idx = (r-1)*X + 1;
    end_idx = r*X;
    RSU_Indices = start_idx:end_idx;
    CurrentSize = AccumulatedSize(r);
    
    % 若未超限，跳过
    if CurrentSize <= C_r
        disp(['RSU ', num2str(r), ' 容量 (', num2str(CurrentSize), ') 满足约束 (', num2str(C_r), ')。']);
        continue;
    end
    disp(['RSU ', num2str(r), ' 容量超限 (', num2str(CurrentSize), ' > ', num2str(C_r), ')。逐次剔除低效用瓦片...']);
    
    % 提取当前RSU内被缓存的瓦片（B_Final=1）
    Cached_Tiles = RSU_Indices(B_Final(RSU_Indices));
    if isempty(Cached_Tiles)
        continue;
    end
    
    % 按效用密度升序排序（先剔除效用最低的）
    [Psi_vals, Sorted_Indices] = sort(Psi(Cached_Tiles), 'ascend');
    Tiles_To_Remove = Cached_Tiles(Sorted_Indices);
    
    % 逐次剔除低效用瓦片，直到容量满足
    remove_count = 0;
    for tile_idx = Tiles_To_Remove
        if AccumulatedSize(r) <= C_r
            break;
        end
        
        % 暂存当前瓦片状态（用于依赖检查）
        tile_size = Tile_Size(tile_idx);
        B_Final(tile_idx) = false; % 剔除当前低效用瓦片
        AccumulatedSize(r) = AccumulatedSize(r) - tile_size;
        remove_count = remove_count + 1;
        
        % 3. 检查并修复依赖连锁剔除：尝试保留同依赖链高效用瓦片
        fix_dependency_violation(B_Final, dependency_matrix, Psi, RSU_Indices, C_r, AccumulatedSize, r, Tile_Size);
    end
    
    disp(['RSU ', num2str(r), ' 剔除 ', num2str(remove_count), ' 个低效用瓦片，最终容量: ', num2str(AccumulatedSize(r))]);
end

% 4. 最终全局依赖校验（仅剔除无法修复的违规）
disp('--- 最终全局依赖校验 ---');
violations_fixed = true;
iteration = 0;
while violations_fixed
    violations_fixed = false;
    tiles_removed = 0;
    iteration = iteration + 1;
    
    for dependent_tile = 1:num_tiles
        if B_Final(dependent_tile) == 1
            Parent_Tiles = find(dependency_matrix(dependent_tile, :) == 1);
            for parent_tile = Parent_Tiles
                if B_Final(parent_tile) == 0
                    % 尝试替换父瓦片：找同RSU内次高效用的父瓦片替代
                    r_parent = ceil(parent_tile / X);
                    parent_rsu_start = (r_parent-1)*X + 1;
                    parent_rsu_end = r_parent*X;
                    parent_candidates = find(B_Final(parent_rsu_start:parent_rsu_end) == 0);
                    parent_candidates = parent_candidates + parent_rsu_start - 1;
                    
                    if ~isempty(parent_candidates)
                        % 按效用密度降序选替代父瓦片
                        [~, sorted_idx] = sort(Psi(parent_candidates), 'descend');
                        best_parent = parent_candidates(sorted_idx(1));
                        % 检查容量是否允许添加
                        if sum(B_Final(parent_rsu_start:parent_rsu_end) .* Tile_Size(parent_rsu_start:parent_rsu_end)) + Tile_Size(best_parent) <= C_RSU(r_parent)
                            B_Final(best_parent) = true;
                            AccumulatedSize(r_parent) = AccumulatedSize(r_parent) + Tile_Size(best_parent);
                            violations_fixed = true;
                            disp(['替换父瓦片：RSU', num2str(r_parent), ' 保留 ', num2str(best_parent), ' 修复 ', num2str(dependent_tile), ' 的依赖']);
                            break;
                        end
                    end
                    
                    % 无替代父瓦片，剔除子瓦片
                    B_Final(dependent_tile) = false;
                    r_dep = ceil(dependent_tile / X);
                    AccumulatedSize(r_dep) = AccumulatedSize(r_dep) - Tile_Size(dependent_tile);
                    violations_fixed = true;
                    tiles_removed = tiles_removed + 1;
                    break;
                end
            end
        end
    end
    
    if tiles_removed > 0
        disp(['第 ', num2str(iteration), ' 轮依赖修正：移除 ', num2str(tiles_removed), ' 个违规子瓦片']);
    end
end

% 输出最终容量
for r = 1:E
    FinalSize = sum(B_Final((r-1)*X + 1 : r*X) .* Tile_Size((r-1)*X + 1 : r*X));
    disp(['最终 RSU ', num2str(r), ' 容量: ', num2str(FinalSize), ' (上限: ', num2str(C_RSU(r)), ')']);
end

end

% 内部辅助函数：修复依赖违规（避免链式剔除）
function fix_dependency_violation(B_Final, dependency_matrix, Psi, RSU_Indices, C_r, AccumulatedSize, r, Tile_Size)
% 检查当前RSU内因剔除瓦片导致的依赖违规，尝试用高效用瓦片替代
violating_tiles = [];
for tile = RSU_Indices
    if B_Final(tile) == 1
        parents = find(dependency_matrix(tile, :) == 1);
        for p = parents
            if B_Final(p) == 0
                violating_tiles = [violating_tiles, tile];
                break;
            end
        end
    end
end

if isempty(violating_tiles)
    return;
end

% 对违规瓦片，尝试找同依赖链的高效用父瓦片
for v_tile = violating_tiles
    parents = find(dependency_matrix(v_tile, :) == 1);
    % 找当前RSU内未被缓存的父瓦片候选
    parent_candidates = intersect(parents, RSU_Indices);
    parent_candidates = parent_candidates(B_Final(parent_candidates) == 0);
    
    if ~isempty(parent_candidates)
        % 按效用密度降序选最优父瓦片
        [~, sorted_idx] = sort(Psi(parent_candidates), 'descend');
        best_parent = parent_candidates(sorted_idx(1));
        % 检查容量是否允许添加
        if AccumulatedSize(r) + Tile_Size(best_parent) <= C_r
            B_Final(best_parent) = true;
            AccumulatedSize(r) = AccumulatedSize(r) + Tile_Size(best_parent);
            disp(['RSU', num2str(r), ' 保留父瓦片 ', num2str(best_parent), ' 修复 ', num2str(v_tile), ' 的依赖']);
        end
    end
end
end

File: algorithm/capacity_refinement_multi_rsu.py
"""
Capacity_Refinement_MultiRSU - Multi-RSU capacity refinement with
dependency-aware tile retention.

Optimized version: removes low-utility tiles one by one instead of
clearing all and re-selecting. When dependency violations occur,
tries to find a higher-utility replacement from the same dependency
chain before falling back to removing the orphan tile.

Input:
    B_MWC: 1-D bool array, MWC cache decisions (length n)
    W_net: 1-D float array, net expected utility
    Tile_Size: 1-D float array, tile sizes
    dependency_matrix: (n, n) dependency matrix
    C_RSU: 1-D float array, capacity per RSU (length E)
    E: number of RSU regions
    X: tiles per RSU region

Output:
    B_Final: 1-D bool array, refined cache decisions
"""

import numpy as np


def _fix_dependency_violation(B_Final, dependency_matrix, Psi,
                              RSU_Indices, C_r, AccumulatedSize, r,
                              Tile_Size):
    """
    Check for dependency violations caused by tile removal within an RSU,
    and attempt to fix by adding high-utility parent tiles.

    Modifies B_Final and AccumulatedSize in-place.
    """
    violating_tiles = []
    for tile in RSU_Indices:
        if B_Final[tile]:
            parents = np.where(dependency_matrix[tile, :] == 1)[0]
            for p in parents:
                if not B_Final[p]:
                    violating_tiles.append(tile)
                    break

    if not violating_tiles:
        return

    # For each violating tile, try to find a high-utility parent
    for v_tile in violating_tiles:
        parents = np.where(dependency_matrix[v_tile, :] == 1)[0]
        # Only consider parents within same RSU
        parent_candidates = np.intersect1d(parents, RSU_Indices)
        # Only parents not already cached
        parent_candidates = parent_candidates[~B_Final[parent_candidates]]

        if len(parent_candidates) > 0:
            # Select the highest-utility-density parent
            best_parent = parent_candidates[np.argmax(Psi[parent_candidates])]
            # Check if adding this parent fits within capacity
            if AccumulatedSize[r] + Tile_Size[best_parent] <= C_r:
                B_Final[best_parent] = True
                AccumulatedSize[r] += Tile_Size[best_parent]
                print(f"  RSU{r+1} 保留父瓦片 {best_parent+1} "
                      f"修复 {v_tile+1} 的依赖")


def capacity_refinement_multi_rsu(B_MWC, W_net, Tile_Size,
                                  dependency_matrix, C_RSU, E, X):
    """
    Refine cache decisions to satisfy per-RSU capacity constraints.

    Parameters
    ----------
    B_MWC : np.ndarray
        1-D bool array, initial MWC cache decisions (length E*X).
    W_net : np.ndarray
        1-D float array, net expected utility per tile.
    Tile_Size : np.ndarray
        1-D float array, size of each tile.
    dependency_matrix : np.ndarray
        (n, n) dependency matrix.
    C_RSU : np.ndarray
        1-D array, capacity limit per RSU (length E).
    E : int
        Number of RSU regions.
    X : int
        Tiles per RSU region.

    Returns
    -------
    B_Final : np.ndarray
        1-D bool array, refined cache decisions.
    """
    num_tiles = len(W_net)
    B_Final = B_MWC.copy()  # preserve initial MWC cache set
    Psi = W_net / Tile_Size  # utility density (higher = keep priority)

    # ---- 1. Initialize per-RSU current sizes ----
    AccumulatedSize = np.zeros(E)
    for r in range(E):
        start_idx = r * X
        end_idx = (r + 1) * X
        AccumulatedSize[r] = np.sum(
            B_Final[start_idx:end_idx] * Tile_Size[start_idx:end_idx]
        )

    # ---- 2. Per-RSU capacity refinement: remove lowest-utility tiles ----
    for r in range(E):
        C_r = C_RSU[r]
        start_idx = r * X
        end_idx = (r + 1) * X
        RSU_Indices = np.arange(start_idx, end_idx)
        CurrentSize = AccumulatedSize[r]

        if CurrentSize <= C_r:
            print(f"  RSU {r+1} 容量 ({CurrentSize:.0f}) "
                  f"满足约束 ({C_r:.0f})。")
            continue

        print(f"  RSU {r+1} 容量超限 ({CurrentSize:.0f} > {C_r:.0f})。"
              f" 逐次剔除低效用瓦片...")

        # Get cached tiles within this RSU
        Cached_Tiles = RSU_Indices[B_Final[RSU_Indices]]
        if len(Cached_Tiles) == 0:
            continue

        # Sort by utility density ascending (remove lowest first)
        psi_vals_order = np.argsort(Psi[Cached_Tiles])  # ascending
        Tiles_To_Remove = Cached_Tiles[psi_vals_order]

        remove_count = 0
        for tile_idx in Tiles_To_Remove:
            if AccumulatedSize[r] <= C_r:
                break

            tile_size = Tile_Size[tile_idx]
            B_Final[tile_idx] = False  # remove this low-utility tile
            AccumulatedSize[r] -= tile_size
            remove_count += 1

            # Check and fix dependency chain violations
            _fix_dependency_violation(
                B_Final, dependency_matrix, Psi, RSU_Indices,
                C_r, AccumulatedSize, r, Tile_Size
            )

        print(f"  RSU {r+1} 剔除 {remove_count} 个低效用瓦片，"
              f"最终容量: {AccumulatedSize[r]:.0f}")

    # ---- 3. Final global dependency validation ----
    print("--- 最终全局依赖校验 ---")
    iteration = 0
    while True:
        iteration += 1
        tiles_removed = 0
        violations_found = False

        for dependent_tile in range(num_tiles):
            if not B_Final[dependent_tile]:
                continue

            Parent_Tiles = np.where(
                dependency_matrix[dependent_tile, :] == 1
            )[0]

            for parent_tile in Parent_Tiles:
                if B_Final[parent_tile]:
                    continue  # parent is cached, OK

                # Try to replace parent: find a high-utility alternative
                # within the same RSU
                r_parent = parent_tile // X
                parent_rsu_start = r_parent * X
                parent_rsu_end = (r_parent + 1) * X

                # Candidate parents: not cached, within same RSU
                parent_candidates = np.arange(parent_rsu_start,
                                              parent_rsu_end)
                parent_candidates = parent_candidates[
                    ~B_Final[parent_candidates]
                ]

                if len(parent_candidates) > 0:
                    best_parent = parent_candidates[
                        np.argmax(Psi[parent_candidates])
                    ]
                    # Check if adding parent fits within capacity
                    current_rsu_size = np.sum(
                        B_Final[parent_rsu_start:parent_rsu_end]
                        * Tile_Size[parent_rsu_start:parent_rsu_end]
                    )
                    if (current_rsu_size + Tile_Size[best_parent]
                            <= C_RSU[r_parent]):
                        B_Final[best_parent] = True
                        AccumulatedSize[r_parent] += Tile_Size[best_parent]
                        print(f"  替换父瓦片：RSU{r_parent+1} "
                              f"保留 {best_parent+1} "
                              f"修复 {dependent_tile+1} 的依赖")
                        violations_found = True
                        break  # move to next dependent_tile

                # No alternative parent → remove orphan child tile
                B_Final[dependent_tile] = False
                r_dep = dependent_tile // X
                AccumulatedSize[r_dep] -= Tile_Size[dependent_tile]
                tiles_removed += 1
                violations_found = True
                break  # move to next dependent_tile

        if tiles_removed > 0:
            print(f"  第 {iteration} 轮依赖修正："
                  f"移除 {tiles_removed} 个违规子瓦片")

        if not violations_found:
            break

    # ---- 4. Print final per-RSU capacity ----
    for r in range(E):
        FinalSize = np.sum(
            B_Final[r * X:(r + 1) * X] * Tile_Size[r * X:(r + 1) * X]
        )
        print(f"  最终 RSU {r+1} 容量: {FinalSize:.0f} "
              f"(上限: {C_RSU[r]:.0f})")

    return B_Final
